src/BehaviorsSDKManaged/Microsoft.Xaml.Interactions.Shared/Media/PlaySoundAction.cs

```
@@ -100,48 +112,67 @@ public object Execute(object sender, object parameter)
-			mediaElement.Volume = this.Volume;
 
-			mediaElement.MediaEnded += this.MediaElement_MediaEnded;
-			mediaElement.MediaFailed += this.MediaElement_MediaFailed;
+            mediaElement.Source = MediaSource.CreateFromUri(sourceUri);
+            mediaElement.AutoPlay = true;
+            mediaElement.MediaPlayer.Volume = this.Volume;
+            mediaElement.MediaPlayer.MediaEnded += this.MediaElement_MediaEnded;
+            mediaElement.MediaPlayer.MediaFailed += this.MediaPlayer_MediaFailed;
 
             this._popup.IsOpen = true;
 			return true;
 		}
 
-		private void MediaElement_MediaFailed(object sender, ExceptionRoutedEventArgs e)
+        private void MediaPlayer_MediaFailed(MediaPlayer sender, MediaPlayerFailedEventArgs args)
 		{
-			if (this._popup != null)
-			{
-				this._popup.IsOpen = false;
-				this._popup.Child = null;
-				this._popup = null;
-			}
+            // TODO: We should probably have some system/properties to report/bubble errors here
+            ClosePopup();
 		}
 
-		private void MediaElement_MediaEnded(object sender, RoutedEventArgs e)
-		{
-			if (this._popup != null)
-			{
-				this._popup.IsOpen = false;
-				this._popup.Child = null;
-				this._popup = null;
-			}
+        private void MediaElement_MediaEnded(MediaPlayer sender, object args)
+        {
+            ClosePopup();
 		}
+
+        private void ClosePopup()
+        {
+            void closePopupImpl()
+            {
+                if (this._popup != null)
+                {
+                    this._popup.IsOpen = false;
+                    this._popup.Child = null;
+                    this._popup = null;
+                }
+            }
+
+            if (_queue.HasThreadAccess)
+            {
+                closePopupImpl();
+            }
+            else
+            {
+                // In WinUI3 the Media events are called on a background thread, so ensure we're on the UI thread to modify our popup container.
```

Does this apply to UWP as well? If so the comment should mention it too for clarity.

---

I'm not sure, I first encountered the issue with WinUI 3 and its MediaPlayerElement, and then migrated UWP to just use the same code path. Since I was building the nuget package to test in my separate test app, I didn't go back and try and bifurcate the code-path here to see the UWP behavior with MediaPlayerElement.

---

Gotcha. Yeah we can leave it like this, don't see any harm anyway 👍

---

Yeah, that's what I figured in case anything ever changed underneath, it's an extra precaution. It could have been silently failing or something too, I just happened to catch it as I had turned exceptions on to figure out what else was going on initially with the migration to WinUI 3.